local_planner/src/nodes/waypoint_generator.cpp

```diff
@@ -191,67 +179,77 @@ bool WaypointGenerator::withinGoalRadius() {
-  float goal_acceptance_radius = 1.0f;
-  if ((goal_ - toEigen(pose_.pose.position)).norm() < goal_acceptance_radius &&
-      std::isfinite(curr_yaw_)) {
-    new_yaw_ = curr_yaw_;
-  }
-
-  // constrain speed
-  Eigen::Vector3f pose_to_wp =
-      toEigen(output_.goto_position) - toEigen(pose_.pose.position);
-
-  if (pose_to_wp.norm() > 0.01f) pose_to_wp.normalize();
+  // goto_position is a unit vector pointing straight up from current location
+  output_.goto_position = pose_.pose.position;
+  output_.goto_position.z += 1.0f;
 
-  pose_to_wp *= planner_info_.min_speed;
-
-  output_.goto_position = toPoint(toEigen(pose_.pose.position) + pose_to_wp);
+  // Will be overwritten by adaptSpeed()
   output_.adapted_goto_position = output_.goto_position;
+
+  // Will be overwritten by smoothWaypoint()
   output_.smoothed_goto_position = output_.goto_position;
-  smoothed_goto_location_ = toEigen(output_.smoothed_goto_position);
 }
 
 void WaypointGenerator::smoothWaypoint(float dt) {
-  if (smoothing_speed_ > 0.01f) {
-    const float P_constant = smoothing_speed_;
-    const float D_constant = 2.f * std::sqrt(P_constant);  // critically damped
-    const Eigen::Vector3f desired_location =
-        toEigen(output_.adapted_goto_position);
-    const Eigen::Vector3f desired_velocity = toEigen(curr_vel_.twist.linear);
-
-    Eigen::Vector3f location_diff = desired_location - smoothed_goto_location_;
-    if (!location_diff.allFinite()) {
-      location_diff = Eigen::Vector3f::Zero();
-    }
-
-    Eigen::Vector3f velocity_diff =
-        desired_velocity - smoothed_goto_location_velocity_;
-    if (!velocity_diff.allFinite()) {
-      velocity_diff = Eigen::Vector3f::Zero();
-    }
+  // In first iteration (very small or invalid dt), initialize the smoothing
+  // to start at the current drone location
+  if (dt <= 0.01) {
+    smoothed_goto_location_ = toEigen(pose_.pose.position);
+    output_.smoothed_goto_position =
+        pose_.pose.position;  // needs to be local position!
+    return;
+  }
 
-    const Eigen::Vector3f p = location_diff * P_constant;
-    const Eigen::Vector3f d = velocity_diff * D_constant;
+  // If the smoothing speed is set to zero, dont smooth, aka use adapted
+  // waypoint
+  // directly
+  if (smoothing_speed_xy_ < 0.01 || smoothing_speed_xy_ < 0.01) {
+    output_.smoothed_goto_position = output_.adapted_goto_position;
+    return;
+  }
 
-    smoothed_goto_location_velocity_ += (p + d) * dt;
-    smoothed_goto_location_ += smoothed_goto_location_velocity_ * dt;
+  // Smooth differently in xz than in z
+  const float P_constant_xy = smoothing_speed_xy_;
+  const float P_constant_z = smoothing_speed_z_;
+  const float D_constant_xy =
+      2.f * std::sqrt(P_constant_xy);  // critically damped
+  const float D_constant_z = 2.f * std::sqrt(P_constant_z);
+  const Eigen::Vector3f P_constant(P_constant_xy, P_constant_xy, P_constant_z);
+  const Eigen::Vector3f D_constant(D_constant_xy, D_constant_xy, D_constant_z);
```

Comment:
Rather calculate D_constant from P_constant with `2*P_constant.array().sqrt()`

Maybe even easier, P_constant and D_constant could be Eigen::Array3f then you wouldn't need all the `array()` calls.

Reply:
Good idea. I can actually get rid of the P_constant_xy floats altogether, and this whole block collapses to two lines!
```
  const Eigen::Array3f P_constant(smoothing_speed_xy_, smoothing_speed_xy_, smoothing_speed_z_);
  const Eigen::Array3f D_constant = 2*P_constant.sqrt();
```